File: ruby/lrucache.rb
class KVPair
  attr_reader :key
  attr_accessor :value
  def initialize(key, value)
    @key, @value = key, value
  end
end

class Node
  attr_accessor :next, :prev, :data
  def initialize(data)
    @data = data
    @prev = nil
    @next = nil
  end
end

class DoublyLinkedList
  attr_reader :len
  def initialize
    @root = Node.new(nil)
    @len = 0

    root.next = root
    root.prev = root
  end

  def unshift(data)
    move_front(Node.new(data)) if inc_len
  end

  def move_front(node)
    return nil if node.nil?
    isolate(node) unless node.prev.nil? && node.next.nil?

    node.prev, node.next = root, root.next
    root.next.prev, root.next = node, node

    node
  end

  def remove_tail
    remove(root.prev)
  end

  def remove(node)
    return nil if len.zero?
    isolate(node) if dec_len
  end

  def isolate(node)
    node.prev.next = node.next
    node.next.prev = node.prev
    node.next, node.prev = nil, nil

    node
  end

  private
  attr_reader :root

  def inc_len; @len += 1; end
  def dec_len; @len -= 1; end
end

class LRUCache
  def initialize(max_size)
    raise 'Max size must be larger than zero' if max_size <= 0
    @max_size = max_size
    @list = DoublyLinkedList.new
    @nodes = Hash.new
  end

  def get(key)
    nodes[key].data.value if list.move_front(nodes[key])
  end

  def set(key, value)
    node = nodes[key]
    if !node.nil?
      node.data.value = value
      list.move_front(node)
      return
    end

    if list.len == max_size
      expired_node = list.remove_tail
      nodes.delete(expired_node.data.key)
    end

    nodes[key] = list.unshift(KVPair.new(key, value))
    nil
  end

  def invalidate(key)
    return nil if nodes[key].nil?

    node = list.remove(nodes[key])
    nodes.delete(key)

    node.data.value
  end

  private
  attr_reader :list, :nodes, :max_size
end

File: javascript/lrucache.js
class KVPair {
  constructor(key, value) {
    this.key = key
    this.value = value
  }
}

class Node {
  constructor(data) {
    this.data = data
    this.next = null
    this.prev = null
  }
}

class DoublyLinkedList {
  constructor() {
    this.root = new Node(null)
    this.len = 0

    this.root.next = this.root
    this.root.prev = this.root
  }

  moveFront(node) {
    if (node === undefined || node === null) {
      return null
    } else if (node.prev !== null && node.next !== null) {
      this.isolate(node)
    }
    const oldHead = this.root.next

    node.prev = this.root
    node.next = oldHead
    oldHead.prev = node
    this.root.next = node

    return node
  }

  unshift(data) {
		const node = new Node(data)
    this.moveFront(node)
    this.incLen()
		return node
  }

  isolate(node) {
    node.prev.next = node.next
    node.next.prev = node.prev
    node.next = null
    node.prev = null
    return node
  }

  remove(node) {
    if (this.len === 0) { return null }
    this.decLen()

    return this.isolate(node)
  }

  removeTail() {
    return this.remove(this.root.prev)
  }

  incLen() { this.len++ }
  decLen() { this.len-- }
}

class LRUCache {
  constructor(maxSize) {
    if (maxSize <= 0) { throw new Error('Max size must be larger than zero') }
    this.maxSize = maxSize
    this.list = new DoublyLinkedList()
    this.nodes = {}
  }

  get(key) {
    const node = this.nodes[key]
    if (node === undefined) {
      return null
    }

    this.list.moveFront(node)
    return node.data.value
  }

  set(key, value) {
    const node = this.nodes[key]
    if (node !== undefined) {
			node.data.value = value
			this.list.moveFront(node)
      return
    }

    if (this.list.len === this.maxSize) {
      const expiredNode = this.list.removeTail()
      delete this.nodes[expiredNode.data.key]
    }

    this.nodes[key] = this.list.unshift(new KVPair(key, value))
  }

  invalidate(key) {
    if (this.nodes[key] === undefined) { return null }

    const node = this.list.remove(this.nodes[key])
    delete this.nodes[key]

    return node.data.value
  }
}

module.exports = LRUCache
